File: src/main/java/org/lanternpowered/server/data/value/mutable/LanternBoundedValue.java
package org.lanternpowered.server.data.value.mutable;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;

import org.lanternpowered.server.data.value.immutable.ImmutableLanternBoundedValue;
import org.spongepowered.api.data.key.Key;
import org.spongepowered.api.data.value.BaseValue;
import org.spongepowered.api.data.value.immutable.ImmutableBoundedValue;
import org.spongepowered.api.data.value.mutable.MutableBoundedValue;

import java.util.Comparator;
import java.util.function.Function;

public class LanternBoundedValue<E> extends LanternValue<E> implements MutableBoundedValue<E> {

    private final Comparator<E> comparator;
    private final E minimum;
    private final E maximum;

    public LanternBoundedValue(Key<? extends BaseValue<E>> key, E defaultValue, Comparator<E> comparator, E minimum, E maximum) {
        this(key, defaultValue, comparator, minimum, maximum, defaultValue);
    }

    public LanternBoundedValue(Key<? extends BaseValue<E>> key, E defaultValue, Comparator<E> comparator, E minimum, E maximum, E actualValue) {
        super(key, defaultValue, actualValue);
        this.comparator = checkNotNull(comparator);
        this.minimum = checkNotNull(minimum);
        this.maximum = checkNotNull(maximum);
        checkState(comparator.compare(maximum, minimum) >= 0);
    }

    @Override
    public E getMinValue() {
        return this.minimum;
    }

    @Override
    public E getMaxValue() {
        return this.maximum;
    }

    @Override
    public Comparator<E> getComparator() {
        return this.comparator;
    }

    @Override
    public MutableBoundedValue<E> set(E value) {
        if (this.comparator.compare(value, this.minimum) >= 0 && this.comparator.compare(value, this.maximum) <= 0) {
            this.actualValue = checkNotNull(value);
        }
        return this;
    }

    @Override
    public MutableBoundedValue<E> transform(Function<E, E> function) {
        return set(checkNotNull(checkNotNull(function).apply(get())));
    }

    @Override
    public ImmutableBoundedValue<E> asImmutable() {
        return new ImmutableLanternBoundedValue<E>(getKey(), this.actualValue, getDefault(), this.comparator, this.minimum, this.maximum);
    }
}

File: src/main/kotlin/org/lanternpowered/server/data/value/LanternBoundedValue.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.data.value

import org.lanternpowered.api.util.uncheckedCast
import org.lanternpowered.api.util.ToStringHelper
import org.lanternpowered.server.data.key.BoundedValueKey
import org.spongepowered.api.data.Key
import org.spongepowered.api.data.value.BoundedValue
import java.util.*

abstract class LanternBoundedValue<E : Any> protected constructor(
        key: Key<out BoundedValue<E>>, value: E, protected val min: () -> E, protected val max: () -> E
) : LanternValue<E>(key, value), BoundedValue<E> {

    override fun getKey() = super.getKey().uncheckedCast<BoundedValueKey<out BoundedValue<E>, E>>()

    override fun getMinValue() = this.min()
    override fun getMaxValue() = this.max()
    override fun getComparator() = this.key.elementComparator

    override fun hashCode() = Objects.hash(this.key, this.value, this.min, this.max, this.comparator)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || this.javaClass != other.javaClass) {
            return false
        }
        other as LanternBoundedValue<*>?
        return this.key == other.key
                && this.value == other.value
                && this.min == other.min
                && this.max == other.max
                && this.comparator == other.comparator
    }

    override fun toString() = ToStringHelper(this)
            .add("key", this.key)
            .add("value", this.value)
            .add("min", this.min)
            .add("max", this.max)
            .add("comparator", this.comparator)
            .toString()
}
